### PR TITLE
expression: add builtin json_keys

### PR DESCRIPTION
**Pull request by Kingwl**
### What problem does this PR solve? 
a part of #7546

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

### CODE REVIEW

File: types/json/binary.go [hunk 1]
```diff
@@ -169,6 +169,16 @@ func (bj BinaryJSON) GetString() []byte {
 	return bj.Value[lenLen : lenLen+int(strLen)]
 }
 
+// GetKeys gets the keys of the object
+func (bj BinaryJSON) GetKeys() BinaryJSON {
+	count := bj.getElemCount()
+	ret := make([]BinaryJSON, 0, count-1)
```
Comment (zz-jason) [inline types/json/binary.go]: should it be `count` rather than `count-1`?

Comment (Kingwl) [reply types/json/binary.go]: sure🌚

File: expression/builtin_json.go [hunk 2]
```diff
@@ -764,8 +765,61 @@ type jsonKeysFunctionClass struct {
...
+	sig.setPbCode(tipb.ScalarFuncSig_JsonKeysSig)
+	return sig, nil
+}
+
+func (b *builtinJSONKeysSig) evalJSON(row chunk.Row) (res json.BinaryJSON, isNull bool, err error) {
+	res, isNull, err = b.args[0].EvalJSON(b.ctx, row)
+	if isNull || err != nil {
+		return res, isNull, errors.Trace(err)
+	}
+	if res.Type() != "OBJECT" {
```
Comment (winoros) [inline expression/builtin_json.go]: Could we use `Typecode`?

File: expression/builtin_json.go [hunk 3]
```diff
@@ -764,8 +765,61 @@ type jsonKeysFunctionClass struct {
...
+	return sig, nil
+}
+
+func (b *builtinJSONKeysSig) evalJSON(row chunk.Row) (res json.BinaryJSON, isNull bool, err error) {
+	res, isNull, err = b.args[0].EvalJSON(b.ctx, row)
+	if isNull || err != nil {
+		return res, isNull, errors.Trace(err)
+	}
+	if res.TypeCode != json.TypeCodeObject {
+		return res, isNull, json.ErrInvalidJSONData
```
Comment (mccxj) [inline expression/builtin_json.go]: isNull -> true

File: expression/builtin_json.go [hunk 4]
```diff
@@ -764,8 +765,61 @@ type jsonKeysFunctionClass struct {
...
+		}
+		pathExpr, err := json.ParseJSONPathExpr(path)
+		if err != nil {
+			return res, true, errors.Trace(err)
+		}
+		if pathExpr.ContainsAnyAsterisk() {
+			return res, true, json.ErrInvalidJSONPathWildcard
+		}
+		var exists bool
+		res, exists = res.Extract([]json.PathExpression{pathExpr})
```
Comment (mccxj) [inline expression/builtin_json.go]: merge line 812 and 813

**Comment by zz-jason** [comment]
@Kingwl please merge master and resolve conflicts.

File: expression/builtin_json_test.go [hunk 5]
```diff
@@ -423,3 +423,72 @@ func (s *testEvaluatorSuite) TestJSONContainsPath(c *C) {
 		}
 	}
 }
+
+func (s *testEvaluatorSuite) TestJSONKeys(c *C) {
```
[the remark this answers is not in the document]
Comment (Kingwl) [reply expression/builtin_json_test.go]: seems already existed

File: expression/builtin_json.go [hunk 6]
```diff
@@ -766,7 +769,90 @@ type jsonKeysFunctionClass struct {
...
+		argTps = append(argTps, types.ETString)
+	}
+	bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETJson, argTps...)
+	var sig builtinFunc
+	switch len(args) {
+	case 1:
+		sig = &builtinJSONKeysSig{bf}
+	case 2:
+		sig = &builtinJSONKeys2ArgsSig{bf}
+	default:
```
Comment (XuHuaiyu) [inline expression/builtin_json.go]: This would never happen since we checked it in line 772.

File: executor/show.go [hunk 7]
```diff
@@ -351,19 +351,19 @@ func (e *ShowExec) fetchShowIndex() error {
 				subPart = col.Length
 			}
 			e.appendRow([]interface{}{
-				tb.Meta().Name.O,       // Table
```
Comment (XuHuaiyu) [inline executor/show.go]: circle ci failed:
```
gofmt (simplify)
executor/show.go
infoschema/tables.go
parser/misc.go
Makefile:94: recipe for target 'fmt' failed
```
These files should not be modified.

File: expression/builtin_json.go [hunk 8]
```diff
@@ -766,7 +769,90 @@ type jsonKeysFunctionClass struct {
...
+		// Should never happens.
+		return nil, errors.Errorf("JSON_KEYS invalid arg length, expect 1 or 2 but got: %v", len(args))
+	}
+	sig.setPbCode(tipb.ScalarFuncSig_JsonKeysSig)
```
Comment (XuHuaiyu) [inline expression/builtin_json.go]: Seems builtinJSONKeysSig and builtinJSONKeys2ArgsSig should be set different PbCode.

File: expression/builtin_json_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (XuHuaiyu) [reply expression/builtin_json_test.go]: You mean line 526?

File: expression/builtin_json.go (hunk 8 above)
Comment (zz-jason) [reply expression/builtin_json.go]: @Kingwl PTAL this comment.

**Comment by Kingwl** [comment]
blocked by https://github.com/pingcap/tidb/pull/7893 for now

**Comment by ngaut** [comment]
PTAL

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

**Pull request merged**